### PR TITLE
Multihash refactoring

### PR DESCRIPTION
**Pull request by sebastianstolzenberg**

### CODE REVIEW

File: src/Options.cpp [hunk 1]
```diff
@@ -536,7 +530,7 @@ bool Options::parseArg(int key, const char *arg)
         return parseArg(key, p ? strtoull(p, nullptr, 16) : strtoull(arg, nullptr, 10));
     }
 
-    case 4013: { /* --doublehash-thread-mask */
+    case 4013: { /* --multihash-thread-mask */
```
Comment (Bendr0id) [inline src/Options.cpp]: To be backwards compatible to old doublehash param i'll add it here.

Comment (sebastianstolzenberg) [reply src/Options.cpp]: Now, I just map both strings to 4013. This should serve the purpose.

```
{ "doublehash-thread-mask",     1, nullptr, 4013 },
{ "multihash-thread-mask",     1, nullptr, 4013 },
```

File: README.md [hunk 2]
```diff
@@ -111,7 +111,7 @@ xmrigDaemon -o pool.minemonero.pro:5555 -u YOUR_WALLET -p x -k --cc-url=IP_OF_CC
   -k, --keepalive                       send keepalived for prevent timeout (need pool support)
   -r, --retries=N                       number of times to retry before switch to backup server (default: 5)
   -R, --retry-pause=N                   time to pause between retries (default: 5)
-      --doublehash-thread-mask          for av=2/4 only, limits doublehash to given threads (mask), (default: all threads)
+      --multihash-thread-mask          for av=2/4 only, limits multihash to given threads (mask), (default: all threads)
```
[the remark this answers is not in the document]
Comment (sebastianstolzenberg) [reply README.md]: done

File: README.md [hunk 3]
```diff
@@ -146,22 +146,22 @@ Also you can use configuration via config file, default **[config.json](https://
 * `--av=3` Software AES implementation.
 * `--av=4` Lower power mode (double hash) of `3`.
 
-## Doublehash thread Mask (only for low power mode (av=2 and av=4))
-With this option you can limit doublehash to the given threads (mask). This can significantly improve your hashrate by using unused l3 cache. The default is to run av2/av4 mode on all threads.
+## Multihash thread Mask (only for low power mode (av=2 and av=4))
```
[the remark this answers is not in the document]
Comment (sebastianstolzenberg) [reply README.md]: done

File: src/Mem_unix.cpp [hunk 4]
```diff
@@ -36,26 +36,20 @@
...
     size_t scratchPadSize = m_algo == Options::ALGO_CRYPTONIGHT ? MEMORY : MEMORY_LITE;
     for (int i=0; i < m_threads; i++) {
         m_memorySize += sizeof(cryptonight_ctx);
-
-        if (isDoubleHash(i)) {
-            m_memorySize += scratchPadSize*2;
-        } else {
-            m_memorySize += scratchPadSize;
-        }
+        m_memorySize += scratchPadSize * m_hashFactor;
```
[the remark this answers is not in the document]
Comment (sebastianstolzenberg) [reply src/Mem_unix.cpp]: It was already fixed in Mem.cpp. Needed to be copied to the unix and the windows versions.

File: src/Mem_win.cpp [hunk 5]
```diff
@@ -148,19 +148,14 @@ bool Mem::allocate(const Options* options)
...
     size_t scratchPadSize = m_algo == Options::ALGO_CRYPTONIGHT ? MEMORY : MEMORY_LITE;
     for (int i=0; i < m_threads; i++) {
         m_memorySize += sizeof(cryptonight_ctx);
-
-        if (isDoubleHash(i)) {
-            m_memorySize += scratchPadSize*2;
-        } else {
-            m_memorySize += scratchPadSize;
-        }
+        m_memorySize += scratchPadSize * m_hashFactor;
```
[the remark this answers is not in the document]
Comment (sebastianstolzenberg) [reply src/Mem_win.cpp]: done

File: src/Options.h [hunk 6]
```diff
@@ -76,8 +81,10 @@ class Options
     inline const char *ccClientConfigFolder() const { return m_ccClientConfigFolder; }
     inline const char *ccCustomDashboard() const    { return m_ccCustomDashboard == nullptr ? "index.html" : m_ccCustomDashboard; }
     inline const std::vector<Url*> &pools() const   { return m_pools; }
-    inline int algo() const                         { return m_algo; }
-    inline int algoVariant() const                  { return m_algoVariant; }
+    inline Algo algo() const                        { return m_algo; }
+//    inline AlgoVariant algoVariant() const          { return m_algoVariant; }
```
[the remark this answers is not in the document]
Comment (sebastianstolzenberg) [reply src/Options.h]: done

File: src/crypto/CryptoNight.cpp [hunk 7]
```diff
@@ -21,6 +21,8 @@
  *   along with this program. If not, see <http://www.gnu.org/licenses/>.
  */
 
+//#include <iostream>
+//#include <boost/chrono.hpp>
```
[the remark this answers is not in the document]
Comment (sebastianstolzenberg) [reply src/crypto/CryptoNight.cpp]: done

File: README.md [hunk 8]
```diff
@@ -135,33 +136,29 @@ xmrigDaemon -o pool.minemonero.pro:5555 -u YOUR_WALLET -p x -k --cc-url=IP_OF_CC
...
-## Algorithm variations
-* `--av=1` For CPUs with hardware AES.
-* `--av=2` Lower power mode (double hash) of `1`.
-* `--av=3` Software AES implementation.
-* `--av=4` Lower power mode (double hash) of `3`.
-
-## Doublehash thread Mask (only for low power mode (av=2 and av=4))
-With this option you can limit doublehash to the given threads (mask). This can significantly improve your hashrate by using unused l3 cache. The default is to run av2/av4 mode on all threads.
+## Multihash thread Mask (only for multihash-factor > 1)
+With this option you can limit multihash to the given threads (mask). This can significantly improve your hashrate by using unused l3 cache. The default is to run av2/av4 mode on all threads.
```
Comment (Bendr0id) [inline README.md]: Add: Short description of --multihash-factor

Outdated: "The default is to run av2/av4 mode on all threads."

File: src/default_config.json [hunk 9]
```diff
@@ -1,7 +1,7 @@
 {
     "algo": "cryptonight",                      // cryptonight (default) or cryptonight-lite
     "av": 0,                                    // algorithm variation, 0 auto select
-    "doublehash-thread-mask" : null,            // for av=2/4 only, limits doublehash to given threads (mask), mask "0x3" means run doublehash on thread 0 and 1 only (default: all threads)
+    "multihash-thread-mask" : null,            // for av=2/4 only, limits multihash to given threads (mask), mask "0x3" means run multihash on thread 0 and 1 only (default: all threads)
```
Comment (Bendr0id) [inline src/default_config.json]: update with content of config.json

**Pull request merged**